pallets/order-book/src/lib.rs

```diff
@@ -571,87 +559,102 @@ pub mod pallet {
-			buy_amount: T::Balance,
-			account_id: T::AccountId,
+			amount_out: T::Balance,
+			fulfilling_account: T::AccountId,
 		) -> DispatchResult {
 			ensure!(
-				buy_amount >= order.min_fulfillment_amount,
-				Error::<T>::InsufficientOrderSize,
+				amount_out >= order.min_fulfillment_amount_out,
+				Error::<T>::BelowMinFulfillmentAmount,
 			);
 
-			ensure!(
-				T::TradeableAsset::can_hold(order.asset_in_id, &(), &account_id, buy_amount),
-				Error::<T>::InsufficientAssetFunds,
-			);
+			let ratio = match order.ratio {
+				OrderRatio::Market => Self::market_ratio(order.asset_out_id, order.asset_in_id)?,
+				OrderRatio::Custom(ratio) => ratio,
+			};
 
-			let sell_amount = Self::convert_with_ratio(
-				order.asset_in_id,
-				order.asset_out_id,
-				order.max_sell_rate,
-				buy_amount,
-			)?;
-			let remaining_buy_amount = order
-				.buy_amount
-				.checked_sub(&buy_amount)
-				.ok_or(Error::<T>::BuyAmountTooLarge)?;
-			let partial_fulfillment = !remaining_buy_amount.is_zero();
+			let amount_in =
+				Self::convert_with_ratio(order.asset_out_id, order.asset_in_id, ratio, amount_out)?;
 
+			let remaining_amount_out = order
+				.amount_out
+				.checked_sub(&amount_out)
+				.ok_or(Error::<T>::FulfillAmountTooLarge)?;
+
+			let partial_fulfillment = !remaining_amount_out.is_zero();
 			if partial_fulfillment {
-				Self::update_order_with_fulfillment(
-					order.placing_account.clone(),
-					order.order_id,
-					remaining_buy_amount,
-					order.max_sell_rate,
-					remaining_buy_amount.min(order.min_fulfillment_amount),
-				)?;
-			} else {
-				T::TradeableAsset::release(
-					order.asset_out_id,
-					&(),
+				let mut updated_order = order.clone();
+				updated_order.amount_out = remaining_amount_out;
+				updated_order.amount_in = order.amount_in.ensure_add(amount_in)?;
+				updated_order.min_fulfillment_amount_out =
+					remaining_amount_out.min(order.min_fulfillment_amount_out);
+
+				<Orders<T>>::insert(updated_order.order_id, updated_order.clone());
+				<UserOrders<T>>::insert(
 					&order.placing_account,
-					sell_amount,
-					Precision::Exact,
-				)?;
-
+					updated_order.order_id,
+					updated_order,
+				);
+			} else {
 				Self::remove_order(order.order_id)?;
 			}
 
+			T::TradeableAsset::release(
+				order.asset_out_id,
+				&(),
+				&order.placing_account,
+				amount_out,
+				Precision::Exact,
+			)?;
 			T::TradeableAsset::transfer(
-				order.asset_in_id,
-				&account_id,
+				order.asset_out_id,
 				&order.placing_account,
-				buy_amount,
+				&fulfilling_account,
+				amount_out,
 				Preservation::Expendable,
 			)?;
 			T::TradeableAsset::transfer(
-				order.asset_out_id,
+				order.asset_in_id,
+				&fulfilling_account,
 				&order.placing_account,
-				&account_id,
-				sell_amount,
+				amount_in,
 				Preservation::Expendable,
 			)?;
 
 			T::FulfilledOrderHook::notify_status_change(
 				order.order_id,
 				Swap {
-					amount: buy_amount,
+					amount: amount_in,
```

Comment:
Maybe we add the fulfilled `amount_out` here as well such that having `Swap` as status type is not sufficient. AFAICS, the we need the `amount_in` at least for `update_investment`. However, we also need `amount_out` for the pending swap state.

If Foreign Investment didn't store pending swaps and instead pulled these directly from the Orderbook, these changes would require less refactoring. It is what it is. One thing after the other.

Reply:
I fully agree with it 👍🏻